rootfs/etc/nginx/lua/util/dns.lua

```diff
@@ -50,26 +69,27 @@ function _M.resolve(host)
     return { host }
   end
 
-  local answers
-  answers, err = r:query(host, { qtype = r.TYPE_A }, {})
-  if not answers then
-    ngx.log(ngx.ERR, "failed to query the DNS server: " .. tostring(err))
-    return { host }
-  end
+  local dns_errors = {}
 
-  if answers.errcode then
-    ngx.log(ngx.ERR, string.format("server returned error code: %s: %s", answers.errcode, answers.errstr))
-    return { host }
+  local addresses, ttl
+  addresses, ttl, err = resolve_host(host, r, r.TYPE_A)
+  if not addresses then
+    table.insert(dns_errors, tostring(err))
+  elseif #addresses ~= 0 then
+    cache:set(host, addresses, ttl)
+    return addresses
   end
 
-  local addresses, ttl = a_records_and_max_ttl(answers)
-  if #addresses == 0 then
-    ngx.log(ngx.ERR, "no A record resolved")
-    return { host }
+  addresses, ttl, err = resolve_host(host, r, r.TYPE_AAAA)
+  if not addresses then
+    table.insert(dns_errors, tostring(err))
+  elseif #addresses ~= 0 then
```

Comment:
nitpick: this can never be less than 0, so I'd do `#addresses > 0`

Reply:
done